### PR TITLE
click around option during crawl

### LINKED ISSUE #25: Crawl beyond the home page
Currently, we only visit the home page of each domain in the scan. Many sites might have different kinds of trackers on different pages. We could modify the crawler to randomly "click" around on the different first-party links on each site it visits.

### PR DESCRIPTION
**Pull request by ablanathtanalba**
Fixes #25 by identifying clickable links on each domain, then iterating through and clicking on each.

### CODE REVIEW

**Comment by ablanathtanalba** [comment]
Most recent commit now filters all clickable options as first party to the given domain.
Question: is it best to limit the amount of links that can be clicked on?

File: crawler.py [hunk 1]
```diff
@@ -509,6 +514,11 @@ def get_domain(self, domain):
             while True:
                 try:
                     self.driver.get(url)
+
+                    # handle click_around option if that's set
+                    if self.click_around:
+                        self.handle_clicking(url)
```
Comment (ghostwords) [inline crawler.py:520]: Do you want the URL or the domain?

File: crawler.py [hunk 2]
```diff
@@ -756,6 +766,23 @@ def save(self, data, name='results.json'):
                 data, f, indent=2, sort_keys=True, separators=(',', ': '))
         self.logger.info("Saved data to %s", name)
 
+    def handle_clicking(self, domain):
+        # fetch the clickable elements on given domain
+        clickables = self.driver.find_elements_by_tag_name('a') + self.driver.find_elements_by_tag_name('button')
+
+        # strip the domain down so we can catch first party subdomains too
+        pattern = re.compile(r"https?://(www\.)?")
+        stripped_domain = pattern.sub('', domain).strip().strip('/')
```
Comment (ghostwords) [inline crawler.py:775]: You don't need this if you simply pass in the domain (not the URL) in the first place.

File: crawler.py [hunk 3]
```diff
@@ -756,6 +766,23 @@ def save(self, data, name='results.json'):
                 data, f, indent=2, sort_keys=True, separators=(',', ': '))
         self.logger.info("Saved data to %s", name)
 
+    def handle_clicking(self, domain):
+        # fetch the clickable elements on given domain
+        clickables = self.driver.find_elements_by_tag_name('a') + self.driver.find_elements_by_tag_name('button')
```
Comment (ghostwords) [inline crawler.py:771]: Buttons don't have an href, so they never get matched, so there is no reason to select them.

File: crawler.py [hunk 4]
```diff
@@ -756,6 +766,23 @@ def save(self, data, name='results.json'):
...
+
+        # strip the domain down so we can catch first party subdomains too
+        pattern = re.compile(r"https?://(www\.)?")
+        stripped_domain = pattern.sub('', domain).strip().strip('/')
+
+        # iterate through each found first party clickable on domain, try to interact with it
+        for clickable in clickables:
+            try:
+                if stripped_domain in clickable.get_attribute('href'):
```
Comment (ghostwords) [inline crawler.py:780]: You could also do this filtering more directly with `find_element_by_css_selector` with [CSS attribute selectors](https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors). For example, `a[href^='https://google.com/']`: all links pointing to https URLs starting with "google.com/".

Comment (ghostwords) [reply crawler.py:780]: Should make sure we don't click on an essentially identical href to the site you are on. No point in clicking on `https://www.facebook.com/#` or `https://www.facebook.com/` on `facebook.com`, for example. (We do now though.)

Comment (ghostwords) [reply crawler.py:780]: This clicks links in a loop, which doesn't actually visit every link. Only the first one (?) wins.

**Comment by ghostwords** [comment]
I suggest adding print statements (just for development debugging) to see stuff like which links get selected for clicking, and also using the `--no-xvfb` option (I haven't tested it with Docker though, sorry) to actually see what the browser ends up doing.

**Comment by ablanathtanalba** [comment]
> I suggest adding print statements (just for development debugging) to see stuff like which links get selected for clicking, and also using the `--no-xvfb` option (I haven't tested it with Docker though, sorry) to actually see what the browser ends up doing.

I've added some print statements. Testing with `--no-xvfb` actually causes the scan to crash -- I didn't realize it until testing it with that suggestion you made. That error isn't special to this branch, I tested on master and it has the same problem with that build as well.

**Comment by ghostwords** [comment]
Implemented by #60.

**Pull request closed without merging**